### PR TITLE
AST: Abstract away the use of TypeLoc for function and subscript result types

### PR DESCRIPTION
**Pull request by AnthonyLatsis**

### CODE REVIEW

File: lib/Sema/TypeCheckDeclObjC.cpp [hunk 1]
```diff
@@ -928,7 +927,7 @@ bool swift::isRepresentableInObjC(const SubscriptDecl *SD, ObjCReason Reason) {
   if (!IndexResult)
     TypeRange = SD->getIndices()->getSourceRange();
   else
-    TypeRange = SD->getElementTypeLoc().getSourceRange();
+    TypeRange = SD->getElementTypeRepr()->getSourceRange();
```
Comment (CodaFi) [inline lib/Sema/TypeCheckDeclObjC.cpp:930]: This part is technically a behavior change that opens us up to a crash-on-invalid case: `TypeLoc` does a null check and converts that to an empty `SourceRange`.

File: lib/Sema/TypeCheckProtocol.cpp [hunk 2]
```diff
@@ -5125,7 +5127,7 @@ diagnoseMissingAppendInterpolationMethod(NominalTypeDecl *typeDecl) {
         
       case InvalidMethod::Reason::ReturnType:
         C.Diags
-            .diagnose(invalidMethod.method->getBodyResultTypeLoc().getLoc(),
+            .diagnose(invalidMethod.method->getResultTypeRepr()->getLoc(),
```
Comment (CodaFi) [inline lib/Sema/TypeCheckProtocol.cpp:5130]: Same here. Have to be careful this repr isn't null.

**Review by CodaFi — APPROVED** [review]
Overall approach is fine. Noticed some things to patch up, but please go through and audit this with the crash-on-invalid case in mind.

File: lib/Sema/TypeCheckDeclObjC.cpp (hunk 1 above)
Comment (AnthonyLatsis) [reply lib/Sema/TypeCheckDeclObjC.cpp:930]: Yeah, thank you. These should be `getElementTypeRepr()`.

**Review by slavapestov — APPROVED** [review]

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test Windows

**Comment by CodaFi** [comment]
@swift-ci please smoke test Windows

**Comment by CodaFi** [comment]
@swift-ci please smoke test macOS

**Comment by AnthonyLatsis** [comment]
@swift-ci please clean smoke test macOS

**Pull request merged**